### PR TITLE
WPCOM Merge: This merges wpcom changes as of 16th of April, 2015

### PR DESCRIPTION
**Pull request by zinigor**
@scruffian, @stephanethomas, @dmsnell could you be so kind and review the changes from your commits and see if they are OK?

### CODE REVIEW

File: class.json-api-endpoints.php [hunk 1]
```diff
@@ -102,6 +102,13 @@
 	 */
 	var $allow_cross_origin_request = false;
 
+	/**
+	 * @var bool Set to true if the endpoint can recieve unauthorized POST requests
+	 *           If you are thinking of setting it, then discuss it with someone:
+	 *           https://operationapi.wordpress.com/2015/04/07/allowing-unauthenticated-post-requests/
```
Comment (jeherve) [inline class.json-api-endpoints.php]: You'll want to remove that URL, it's a private P2.

**Comment by stephanethomas** [comment]
Well, I can certainly review my changes again but you could provide more context? What are you looking for exactly?

**Comment by zinigor** [comment]
@stephanethomas what I had in mind was just a sanity check - to make sure that your changes here are represented how they were intended and that there's nothing missing. Thanks!

File: class.json-api-endpoints.php [hunk 2]
```diff
@@ -96,12 +96,14 @@
 
 	/**
 	 * @var bool Set to true if the endpoint accepts all cross origin requests
-	 *    You probably should not set this flag. If you are thinking of setting it,
-	 *    then discuss it with someone:
-	 *       http://operationapi.wordpress.com/2014/06/25/patch-allowing-endpoints-to-do-cross-origin-requests/
```
Comment (scruffian) [inline class.json-api-endpoints.php]: I don't think this block should be removed.

**Comment by scruffian** [comment]
I think you are missing changes from `public.api/rest/class.wpcom-json-api.php` - You might need to make equivalent changes somewhere else, otherwise there's no point in having the two new methods (`allows_cross_origin_requests` and `allows_unauthorized_requests`).

(original commit https://wpcom.trac.automattic.com/changeset/116175)

File: class.json-api-endpoints.php (hunk 2 above)
Comment (scruffian) [inline class.json-api-endpoints.php]: @jeherve pointed out we need to remove this reference to a P2.

**Comment by zinigor** [comment]
I'm going to re-sync those changes in order to get the comments properly. Some time later we will need to examine what code we sync so that we don't have unused parts of code in the REST API that rely on unsynced code in wpcom.

File: class.json-api-endpoints.php (hunk 2 above)
Comment (kraftbj) [inline class.json-api-endpoints.php]: Or at least hide it in Jetpack. On the wpcom version, you can use the commenting style mentioned in the "Adding Files to the Build Script" article in the FG to selectively hide comments from JP that would remain on wp.com

File: class.json-api-endpoints.php [hunk 3]
```diff
@@ -1963,6 +1937,14 @@ function is_current_site_multi_user() {
 		return $users > 1;
 	}
 
+	function allows_cross_origin_requests() {
+		return 'GET' == $this->method || $this->allow_cross_origin_request;
+	}
+
+	function allows_unauthorized_requests( $origin, $complete_access_origins  ) {
```
Comment (dereksmart) [inline class.json-api-endpoints.php]: Extra whitespace at the end of line

**Comment by dereksmart** [comment]
LGTM after the minor whitespace fix!

**Pull request merged**